Zend/zend_weakrefs.c

```diff
@@ -0,0 +1,217 @@
+/*
+   +----------------------------------------------------------------------+
+   | PHP Version 7                                                        |
+   +----------------------------------------------------------------------+
+   | Copyright (c) The PHP Group                                          |
+   +----------------------------------------------------------------------+
+   | This source file is subject to version 2.00 of the Zend license,     |
+   | that is bundled with this package in the file LICENSE, and is        |
+   | available through the world-wide-web at the following url:           |
+   | http://www.zend.com/license/2_00.txt.                                |
+   | If you did not receive a copy of the Zend license and are unable to  |
+   | obtain it through the world-wide-web, please send a note to          |
+   | license@zend.com so we can mail you a copy immediately.              |
+   +----------------------------------------------------------------------+
+   | Authors: krakjoe@php.net                                             |
+   +----------------------------------------------------------------------+
+*/
+
+#include "zend.h"
+#include "zend_interfaces.h"
+#include "zend_objects_API.h"
+#include "zend_weakrefs.h"
+
+typedef struct _zend_weakref {
+	zval referent;
+	zend_object std;
+} zend_weakref;
+
+zend_class_entry *zend_ce_weakref;
+static zend_object_handlers zend_weakref_handlers;
+
+#define zend_weakref_from(o) ((zend_weakref*)(((char*) o) - XtOffsetOf(zend_weakref, std)))
+#define zend_weakref_fetch(z) zend_weakref_from(Z_OBJ_P(z))
+
+static void zend_weakref_unref(zval *zv) {
+	zend_weakref *wr = (zend_weakref*) Z_PTR_P(zv);
+
+	GC_DEL_FLAGS(Z_OBJ(wr->referent), IS_OBJ_WEAKLY_REFERENCED);
+
+	ZVAL_NULL(&wr->referent);
+}
+
+void zend_weakrefs_init() {
+	zend_hash_init(&EG(weakrefs), 8, NULL, zend_weakref_unref, 0);
+}
+
+void zend_weakrefs_notify(zend_object *object) {
+	zend_hash_index_del(&EG(weakrefs), (zend_ulong) object);
+}
+
+void zend_weakrefs_shutdown() {
+	zend_hash_destroy(&EG(weakrefs));
+}
+
+static zend_object* zend_weakref_clone(zval *zv) {
+	Z_ADDREF_P(zv);
+
+	return Z_OBJ_P(zv);
```

Comment:
Hm... this looks odd. I'm not sure if it's okay for clone to return the same object.

Reply:
Yeah, I thought that too ... deny clone also feels strange though, it will force people to do

```
$clone = WeakReference::of($wr->get());
```

Which feels just as strange ?

Reply:
That's a very roundabout way of writing `$clone = $wr` :D

Reply:
Dropped clone ...